ipc_receivefile.c

```diff
@@ -0,0 +1,408 @@
+				}
+				if(optarg==NULL){
+					buffer_size = 4;
+				}else{
+					buffer_size = optarg;
+				}
+				method='s';
+				break;
+        //other situation
+        case 'q':
+        case 'h':
+        default:
+    	  		printf ("Usage: my_program [-m] for file transfer by message passing\n");
+    	  		printf ("                  [-p] for file transfer by named pipe\n");
+    	  		printf ("                  [-s] for file transfer by shared memory,");
+    	  		printf ("                  [-f file_path] for choosing the target file\n");
+    	  		printf ("                  [-h] for this help command\n");
+    	  		printf ("                  note: sender and receiver should use the same method!\n");
+                return EXIT_FAILURE;
+        }
+    }
+
+    if(fd==-1){
+    	printf("the file to receive information has not been provided or error happens!");
+    	return EXIT_FAILURE;
+    }
+    switch (method){
+    	case 'm':
+    			ipcMessagePassingReceive(fd);
+    			break;
+    	case 'p':
+    			ipcPipeReceive(fd);
+    			break;
+    	case 's':
+    			ipcSharedMemoryReceive(fd);
+    			break;
+    	default:
+    			return EXIT_FAILURE;
+    }
+	return EXIT_SUCCESS;
+}
+
+
+void ipcMessagePassingReceive(int fd){
+	int rcvid;
+	name_attach_t* attach;
+	msg_buf_t msg;
+	int status;
+	int reply_status;
+	char* data;
+
+	attach = name_attach(NULL, CKSUM_SERVER_NAME, 0);
+	if (attach == NULL)
+	{
+		perror("name_attach");
+		exit(EXIT_FAILURE);
+	}
+
+	while (1)
+	{
+		printf("Waiting for a message...\n");
+		rcvid = MsgReceive(attach->chid, &msg, sizeof(msg), NULL);
+		if (rcvid == -1)
+		{
+			perror("MsgReceive");
+			exit(EXIT_FAILURE);
+		}
+		else if (rcvid > 0)
+		{ //msg
+			switch (msg.msg_type)
+			{
+			case CKSUM_IOV_MSG_TYPE:
+				printf("Received a checksum request msg, header says the data is %d bytes\n",
+						msg.cksum_hdr.data_size);
+				data = malloc(msg.cksum_hdr.data_size);
+				if (data == NULL)
+				{
+					if (MsgError(rcvid, ENOMEM ) == -1)
+					{
+						perror("MsgError");
+						exit(EXIT_FAILURE);
+					}
+				}
+				else
+				{
+					status = MsgRead(rcvid, data, msg.cksum_hdr.data_size, sizeof(cksum_header_t));
+					if (status == -1)
+					{
+						perror("MsgRead");
+						exit(EXIT_FAILURE);
+					}
+
+
+					int write_size = write(fd, data, msg.cksum_hdr.data_size);
+				    if( write_size == -1 ) {
+				        perror( "write error" );
+						exit(EXIT_FAILURE);
+				    }
+
+					free(data);
+
+					status = MsgReply(rcvid, EOK, &reply_status, sizeof(reply_status));
+					if (status == -1)
+					{
+						perror("MsgReply");
+						exit(EXIT_FAILURE);
+					}
+				}
+
+				break;
+			default:
+				if (MsgError(rcvid, ENOSYS) == -1)
+				{
+					perror("MsgError");
+				}
+				break;
+			}
+		}
+		else
+		{
+			printf("Receive returned an unexpected value: %d\n", rcvid);
+		}
+	}
+}
+
+void ipcPipeReceive(int fd){
+    int fifofd;
+    // FIFO file path
+	int read_size=0;
+	int write_size=0;

```

Comment:
write_size is not used

Reply:
Thank you for pointing out. All above has been changed. Could you try it now?